Add Netutils and register all jinja filters into django-jinja

dgarros opened the pull request:
### Fixes: #1082

This PR adds Netutils and register all jinja filters into django-jinja (~53 filters). The complete list of filter available is available [here](https://github.com/networktocode/netutils/blob/develop/netutils/utils.py#L4) 
As it it's functional but before putting more time into it I would like to understand the level of interest around this PR

> Right now I'm registering the filter as part of the extras app, not sure if this is the best place to do that in Nautobot

```python
### Nautobot interactive shell (e2010bc85319)
### Python 3.6.15 | Django 3.1.13 | Nautobot 1.2.0a1
### lsmodels() will show available models. Use help(<model>) for more info.
>>>
>>>
>>> from nautobot.utilities.utils import render_jinja2
>>> render_jinja2("{{ ip | ip_to_hex }}", {"ip": "1.1.1.1" })
'01010101'
>>>
```

### TODO:
- Update documentation
- Add unit tests

itdependsnetworks commented on pyproject.toml:
```diff
@@ -31,7 +31,7 @@ include = [
 
 [tool.poetry.dependencies]
-python = "^3.6"
+python = "^3.6.2"
```
This will be addressed shortly, likely was from a copy/paste from other project https://github.com/networktocode/netutils/pull/75

glennmatthews commented on nautobot/extras/apps.py:
```diff
@@ -44,3 +44,10 @@ def convert_field_to_list_tags(field, registry=None):
+
+        # Register netutils jinja2 filters
+        from netutils.utils import jinja2_convenience_function
+        from django_jinja.library import filter, _register_function
+
+        for name, func in jinja2_convenience_function().items():
+            _register_function("filters", name=name, fn=func)
```
Looks like this could be achieved without using a private function as follows:

```suggestion
        from django_jinja import library

        for name, func in jinja2_convenience_function().items():
            library.filter(name=name, fn=func)
```

glennmatthews commented on nautobot/extras/apps.py:
```diff
@@ -44,3 +44,10 @@ def convert_field_to_list_tags(field, registry=None):
 
         plugin_dir.register(DatabaseBackend)
+
+        # Register netutils jinja2 filters
```
Not clear to me that this needs to be done inside an app `ready()` function. This code might be a better fit somewhere like `nautobot/utilities/utils.py`?

dgarros replied:
I think it's important to register these filters when the Nautobot is starting, as far as I know I don't think they will be loaded automatically in `nautobot/utilities/utils.py`

glennmatthews replied:
I suggested that file as that's where the `render_jinja2` function is defined - any code that consumes these filters should be using that function as its renderer, generally speaking.

glennmatthews commented on pyproject.toml:
```diff
@@ -106,6 +106,8 @@ PyYAML = "~5.4.1"
 social-auth-app-django = "^4.0.0"
 # Rendering of SVG images (for rack elevations, etc.)
 svgwrite = "~1.4.1"
+# Library of network related Jinja Filters
+netutils = "~0.2.4"
```
Please move to alphabetical order among other requirements.

glennmatthews reviewed (commented):
Ideally I'd like to see documentation of the new filters being provided this way (either in nautobot's core docs, or at minimum a reference to netutil's own docs from anywhere Nautobot can use these filters, notably in the computed-fields docs) and at least a few unit tests demonstrating that these filters are usable by Nautobot.

It'd be interesting to see if there are any core filters or template-tags within Nautobot that could be replaced by usage of netutils - offhand I'm aware of `nautobot.utilities.templatetags.helpers.humanize_speed()` as one such candidate but there may very well be others.

glennmatthews commented:
For this to make it into 1.2.0, we'd like to see the following:

- As an NTC-maintained library, we'd like to see Netutils reach a 1.0 version and commit to semantic versioning, so that we can have confidence that minor version number changes in netutils won't introduce breaking changes in these filters that would be user-impacting
- Documentation either in nautobot or otherwise web-accessible of all of these filters and their usage
- A reasonable cross-section of unit tests in Nautobot verifying that these filters are usable in computed fields, export templates, etc.

jedelman8 commented:
@itdependsnetworks what do you think? Will you be taking this on in the next week or so?

itdependsnetworks commented:
yes, will be doing this by next week.

dgarros commented:
I've made several changes to my branch and I think it's ready for review, here is the list of all the changes
- Upgraded netutils to 1.0.0 and move it in pyproject.toml
- Moved registration of filters to utilities `ready()` instead of extras `ready()` since the other functions are already in utilities
- Added unit tests and documentation for netutils filters
- Added unit tests and documentation for built-in helpers
- Changed target branch to `develop` instead of `next`

Also, while working on this PR, 
- I noticed that all built-in Django Template helpers in `templatetags/helpes.py` were not available in Django Jinja so I loaded them on both sides and I loaded all netutils functions on both sides as well to have feature parity between Jinja2 and Django Template.
- I also found couple of bugs in `templatetags/helpes.py` while writing the unit tests so I fixed them directly

Is there an easy way to generate the doc locally to validate that everything is rendering as expected ?

glennmatthews commented:
> Is there an easy way to generate the doc locally to validate that everything is rendering as expected ?

`mkdocs serve` and connect to `http://localhost:8001`.

Thanks for the updates!

glennmatthews commented on nautobot/utilities/templatetags/helpers.py:
```diff
@@ -69,48 +85,86 @@ def render_yaml(value):
+        str: string in Title format
+
+    Example:
+        {{ obj_type.name|bettertitle }}
```
```suggestion
        >>> bettertitle("IP address")
        "IP Address"
```

glennmatthews commented on nautobot/utilities/templatetags/helpers.py:
```diff
@@ -212,41 +306,81 @@ def has_one_or_more_perms(user, permissions_list):
     try:
         int(n)
-    except TypeError:
+    except (TypeError, ValueError):
```
When's a `ValueError` seen here?

dgarros replied:
if n is a string that can't be converted into an int
```
>>> int("one")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid literal for int() with base 10: 'one'
```

glennmatthews commented on nautobot/utilities/tests/test_templatetags_helpers.py:
```diff
@@ -0,0 +1,141 @@
+
+    def test_fgcolor(self):
+        self.assertEqual(fgcolor("#999999"), "#ffffff")
+        self.assertEqual(fgcolor("#111111"), "#ffffff")
```
Would be good to have at least one test that returns `#000000`.

glennmatthews commented:
Thank you for adding documentation and unit tests!

dgarros commented on nautobot/docs/additional-features/jinja-filters.md:
```diff
@@ -0,0 +1,239 @@
+# Additional Jinja2 filters and Django Template helpers
```
Thanks i added a link in few places

dgarros commented on nautobot/docs/additional-features/jinja-filters.md:
````diff
@@ -0,0 +1,239 @@
+
+```no-highlight
+{{ speed_value | humanize_speed }}
+```
````
Good point, at some point I think we could call `bits_to_name` within `humanize_speed` to leverage the code from netutils, we would also need to convert kbps to bits

dgarros commented:
I made the suggested changes, the PR should be good for review

glennmatthews reviewed (approved):
Thank you!
